pkg/server/server.go

```diff
@@ -242,33 +245,68 @@ func (s *Server) Write(ctx context.Context, req *openfgapb.WriteRequest) (*openf
 }
 
 func (s *Server) Check(ctx context.Context, req *openfgapb.CheckRequest) (*openfgapb.CheckResponse, error) {
+
 	store := req.GetStoreId()
 	tk := req.GetTupleKey()
 	ctx, span := s.tracer.Start(ctx, "check", trace.WithAttributes(
-		attribute.KeyValue{Key: "store", Value: attribute.StringValue(req.GetStoreId())},
+		attribute.KeyValue{Key: "store", Value: attribute.StringValue(store)},
 		attribute.KeyValue{Key: "object", Value: attribute.StringValue(tk.GetObject())},
 		attribute.KeyValue{Key: "relation", Value: attribute.StringValue(tk.GetRelation())},
 		attribute.KeyValue{Key: "user", Value: attribute.StringValue(tk.GetUser())},
 	))
 	defer span.End()
 
+	if tk.GetUser() == "" || tk.GetRelation() == "" || tk.GetObject() == "" {
+		return nil, serverErrors.InvalidCheckInput
+	}
+
 	modelID, err := s.resolveAuthorizationModelID(ctx, store, req.GetAuthorizationModelId())
 	if err != nil {
 		return nil, err
 	}
 	span.SetAttributes(attribute.KeyValue{Key: "authorization-model-id", Value: attribute.StringValue(modelID)})
 
-	q := commands.NewCheckQuery(s.datastore, s.tracer, s.meter, s.logger, s.config.ResolveNodeLimit)
+	model, err := s.datastore.ReadAuthorizationModel(ctx, store, modelID)
+	if err != nil {
+		return nil, err
+	}
 
-	res, err := q.Execute(ctx, &openfgapb.CheckRequest{
-		StoreId:              store,
-		TupleKey:             tk,
-		ContextualTuples:     req.GetContextualTuples(),
-		AuthorizationModelId: modelID,
-		Trace:                req.GetTrace(),
+	typesys := typesystem.New(model)
+
+	if err := validation.ValidateUserObjectRelation(typesys, tk); err != nil {
+		return nil, serverErrors.ValidationError(err)
+	}
+
+	for _, ctxTuple := range req.GetContextualTuples().GetTupleKeys() {
+		if err := validation.ValidateTuple(typesys, ctxTuple); err != nil {
+			return nil, serverErrors.HandleTupleValidateError(err)
+		}
+	}
+
+	g := graph.NewConcurrentChecker(storage.NewContextualTupleDatastore(s.datastore), 100)
+
+	ctx = typesystem.ContextWithTypesystem(ctx, typesystem.New(model))
+	ctx = storage.ContextWithContextualTuples(ctx, req.ContextualTuples.GetTupleKeys())
```

Comment:
Notice that ContextualTuples are now truly "contextual" 😄 . We pass contextual tuples in through the context and the storage layer is responsible for extracting any tuples in the context, filtering them based on the query filters, and including/merging the filtered tuples in the returned tuple iterator (which combines the tuples from the database query and the contextual tuples).

Reply:
I'd rather simply remove the `contextualtuples` package in this PR, as we may forget if we don't. Let's just accept this as the way forward and fix it here if there are problems. WDYT?

Reply:
To minimize the scope of this PR, I'd like to save the removal of existing packages for a subsequent PR. But I agree, we should drop the `contextualtuples` package and remove the existing Check implementation once this is merged 👍